ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/forkchoice/TerminalPowBlockMonitorTest.java

```diff
@@ -0,0 +1,270 @@
+  private static final UInt64 MERGE_FORK_EPOCH = UInt64.ONE;
+  private static final UInt256 TTD = UInt256.valueOf(10_000_000);
+  private static final Bytes32 TERMINAL_BLOCK_HASH = Bytes32.random();
+  private static final UInt64 TERMINAL_BLOCK_EPOCH = UInt64.valueOf(2);
+
+  private final ExecutionEngineChannel executionEngine = mock(ExecutionEngineChannel.class);
+  private final StubTimeProvider timeProvider = StubTimeProvider.withTimeInSeconds(10_000);
+  private final StubAsyncRunner asyncRunner = new StubAsyncRunner(timeProvider);
+  private final ForkChoiceNotifier forkChoiceNotifier = mock(ForkChoiceNotifier.class);
+
+  private Spec spec;
+  private DataStructureUtil dataStructureUtil;
+  private StorageSystem storageSystem;
+  private RecentChainData recentChainData;
+  private TerminalPowBlockMonitor terminalPowBlockMonitor;
+
+  @BeforeAll
+  public static void initSession() {
+    AbstractBlockProcessor.BLS_VERIFY_DEPOSIT = false;
+  }
+
+  @AfterAll
+  public static void resetSession() {
+    AbstractBlockProcessor.BLS_VERIFY_DEPOSIT = true;
+  }
+
+  private void setUpTerminalBlockHashConfig() {
+    setUpCommon(
+        mergeBuilder ->
+            mergeBuilder
+                .mergeForkEpoch(MERGE_FORK_EPOCH)
+                .terminalBlockHash(TERMINAL_BLOCK_HASH)
+                .terminalBlockHashActivationEpoch(TERMINAL_BLOCK_EPOCH));
+  }
+
+  private void setUpTTDConfig() {
+    setUpCommon(
+        mergeBuilder -> mergeBuilder.mergeForkEpoch(MERGE_FORK_EPOCH).terminalTotalDifficulty(TTD));
+  }
+
+  private void setUpCommon(Consumer<SpecConfigBuilder.MergeBuilder> mergeBuilder) {
+    spec =
+        TestSpecFactory.createMerge(
+            SpecConfigLoader.loadConfig(
+                "minimal",
+                phase0Builder ->
+                    phase0Builder
+                        .altairBuilder(altairBuilder -> altairBuilder.altairForkEpoch(UInt64.ZERO))
+                        .mergeBuilder(mergeBuilder)));
+    dataStructureUtil = new DataStructureUtil(spec);
+    storageSystem = InMemoryStorageSystemBuilder.buildDefault(spec);
+    storageSystem.chainUpdater().initializeGenesis(false);
+    recentChainData = storageSystem.recentChainData();
+
+    terminalPowBlockMonitor =
+        new TerminalPowBlockMonitor(
+            executionEngine, spec, recentChainData, forkChoiceNotifier, asyncRunner);
+  }
+
+  private void goToSlot(UInt64 slot) {
+    storageSystem
+        .chainUpdater()
+        .updateBestBlock(storageSystem.chainUpdater().advanceChainUntil(slot));
+  }
+
+  private void doMerge(Bytes32 terminalBlockHash) {
+    SignedBlockAndState newBlockWithExecutionPayloadAtopTerminalBlock =
+        storageSystem
+            .chainUpdater()
+            .chainBuilder
+            .generateBlockAtSlot(
+                recentChainData.getHeadSlot().plus(1),
+                ChainBuilder.BlockOptions.create()
+                    .setTransactions(dataStructureUtil.randomBytes(40))
```

Comment:
FYI: You can just use `setTransactions()` - as long as it's called it will generate a non-default payload, you don't actually need a transaction.  Maybe we should add a `withExecutionPayload()` method that makes that a bit more obvious..

Reply:
I approached this a little differently: before merge, if you want to trigger the transition you use `setTerminalBlockHash`. after that, chain builder will always generate a payload on top of the last available.